lib/ansible/plugins/connection/ssh.py

```diff
@@ -578,11 +589,20 @@ def _build_command(self, binary, subsystem, *other_args):
 
         # If we want to use password authentication, we have to set up a pipe to
         # write the password to sshpass.
-
-        if conn_password:
+        pkcs11_provider = self.get_option("pkcs11_provider")
+        if conn_password or len(pkcs11_provider) >= 1:
             if not self._sshpass_available():
-                raise AnsibleError("to use the 'ssh' connection type with passwords, you must install the sshpass program")
+                raise AnsibleError("to use the 'ssh' connection type with passwords or pkcs11_provider, you must install the sshpass program")
 
+        if conn_password and len(pkcs11_provider) >= 1:
+            self.sshpass_pipe = os.pipe()
+            b_command += [b'sshpass',
+                          b'-d' + to_bytes(self.sshpass_pipe[0], nonstring='simplerepr', errors='surrogate_or_strict'),
+                          b'-P',
+                          b'Enter PIN for ']
+        elif conn_password and len(pkcs11_provider) > 0:
+            raise AnsibleError("To use pkcs11_provider you must specify a password/pin")
+        elif conn_password:
```

Comment:
There's a bit of duplication of code here. We essentially do the same as the password and pkcs11_provider check but the prompt is optional/based on `sshpass_prompt`.

It seems to be that we should do something like

```python
if conn_password or pkcs11_provider:
    if pkcs11_provider and not conn_password:
        raise AnsibleError('need password/pin ...')

    self.sshpass_pipe = os.pipe()
    b_command += [b'sshpass', b'-d' + to_bytes(self.sshpass_pipe[0], nonstring='simplerepr', errors='surrogate_or_strict')]

    password_prompt = self.get_options('sshpass_prompt')
    if not password_prompt and pkcs11_provider:
        password_prompt = 'Enter PIN for '

    if password_prompt:
        b_command += [b'-P', to_bytes(password_prompt, ...)]

    ...
```

Now the same code is used when setting up sshpass and we now have a way to override the expected prompt even when using PKCS11. The docs for `sshpass_prompt` should be updated to say that it defaults to `Enter PIN for` when `pkcs11_provider` is set.

Reply:
I just took care of the duplication with commit c254a33.

Reply:
I just added a changelog fragment as well.